### PR TITLE
Users/lukillgo/mavenfeedauth

### PR DESCRIPTION
**Pull request by lkillgore**
Maven feed authentication.

### CODE REVIEW

File: Tasks/Maven/maventask.ts [hunk 1]
```diff
@@ -155,21 +201,24 @@ async function execBuild() {
...
+            mvnRun = codeAnalysisOrchestrator.configureBuild(mvnRun);
 
             // Read Maven standard output
             mvnRun.on('stdout', function (data) {
                 processMavenOutput(data);
             });
 
             // 3. Run Maven. Compilation or test errors will cause this to fail.
-            return mvnRun.exec(); // Run Maven with the user specified goals
+            return mvnRun.execSync(util.getExecOptions());
```
Comment (madhurig) [inline Tasks/Maven/maventask.ts]: execSync() will block right? We should use execSync() for very quick operations only.

File: Tasks/Maven/maventask.ts [hunk 2]
```diff
@@ -147,6 +150,49 @@ async function execBuild() {
...
+                    }
+                }
+                return util.mergeCredentialsIntoSettingsXml(settingsXmlFile, repositories);
+            })
+            .fail(function (err) {
+                return Q.reject(err);
+            });
+        })
+        .fail(function (err) {
+            console.error(err.message);
```
Comment (madhurig) [inline Tasks/Maven/maventask.ts]: tl.error() instead of console.error()

File: Tasks/Maven/mavenutil.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    return writeJsonAsXmlFile(filePath, jsonContent.project, 'project');
+}
+
+function writeFile(filePath: string, fileContent: string): Q.Promise<void> {
+    fse.mkdirpSync(path.dirname(filePath));
+    return Q.nfcall<void>(fs.writeFile, filePath, fileContent, { encoding: 'utf-8' });
+}
+
+function addPropToJson(obj: any, propName:string, value: any): void {
+    if (typeof obj === 'undefined') {
```
Comment (madhurig) [inline Tasks/Maven/mavenutil.ts]: what about null check? if (!obj) might work better.

File: Tasks/Maven/package.json [hunk 4]
```diff
@@ -11,6 +11,12 @@
...
     "request": "^2.74.0",
     "vsts-task-lib": "1.1.0",
-    "xml2js": "^0.4.16"
+    "xml2js": "^0.4.16",
+    "base-64": "^0.1.0",
+    "utf8": "^2.1.0",
+    "cheerio": "^0.22.0",
+    "strip-bom": "^3.0.0",
+    "fs-extra": "^0.30.0",
+    "vso-node-api": "6.0.1-preview"
```
Comment (madhurig) [inline Tasks/Maven/package.json]: Double check that all of these are approved for use

Comment (lkillgore) [inline Tasks/Maven/package.json]: Yes, these are used by other modules in vsts.

File: Tasks/Maven/mavenutil.ts (hunk 3 above)
Comment (lkillgore) [inline Tasks/Maven/mavenutil.ts]: Done

File: Tasks/Maven/maventask.ts (hunk 1 above)
Comment (lkillgore) [inline Tasks/Maven/maventask.ts]: Good catch.  I'm not sure how this change made it through.

File: Tasks/Maven/maventask.ts (hunk 2 above)
Comment (lkillgore) [inline Tasks/Maven/maventask.ts]: I'm inclined to leave this one alone.  The rest of this module uses console.error (and I copied its use).  I'm thinking change them all or be consistent.

Comment (lkillgore) [inline Tasks/Maven/maventask.ts]: I recant my objection.  There are other lines that use 'tl.error', so there's no sense of consistency anyway.

File: Tasks/Maven/maventask.ts [hunk 5]
```diff
@@ -153,6 +153,10 @@ async function execBuild() {
             mvnRun.arg('help:effective-pom');
             return util.collectFeedRepositoriesFromEffectivePom(mvnRun.execSync()['stdout'])
             .then(function (repositories) {
+                if (!repositories || !repositories.length) {
```
Comment (madhurig) [inline Tasks/Maven/maventask.ts]: Do we need to check for repositories.length===0?

File: Tests-Legacy/L0/Maven/_suite.ts [hunk 6]
```diff
@@ -549,6 +549,35 @@ describe('Maven Suite', function () {
             });
     })
 
+    it('run maven with feed', (done) => {
```
Comment (madhurig) [inline Tests-Legacy/L0/Maven/_suite.ts]: Will be good to add a test for "run maven with feed when user specifies settings file" as well.

File: Tasks/Maven/maventask.ts (hunk 5 above)
Comment (lkillgore) [inline Tasks/Maven/maventask.ts]: !repositories.length should work for that purpose, no?

Comment (madhurig) [inline Tasks/Maven/maventask.ts]: You are right, I prefer explicit checks but this is fine.

**Pull request merged**